src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,182 @@
+                                                                                      isDataflowRule: false);
+
+        private const string RequiresPreviewFeaturesAttribute = nameof(RequiresPreviewFeaturesAttribute);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        private bool thisAssemblyUsesPreviewFeatures;
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context =>
+            {
+                IAssemblySymbol? thisAssembly = context.Compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass.ToString() == RequiresPreviewFeaturesAttribute);
+                if (!(attribute is null))
+                {
+                    // This assembly has enabled preview attributes.
+                    return;
+                }
+
+                var enabledPreviewFeatures = context.Options.GetMSBuildPropertyValue(MSBuildPropertyOptionNames.EnablePreviewFeatures, context.Compilation, context.CancellationToken);
+                if (string.Equals(enabledPreviewFeatures, "true", System.StringComparison.OrdinalIgnoreCase))
+                {
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, IOperation?>();
+                context.RegisterOperationBlockStartAction(context => BuildSymbolInformation(context, requiresPreviewFeaturesSymbols));
+            });
+        }
+
+        private void BuildSymbolInformation(OperationBlockStartAnalysisContext context, ConcurrentDictionary<ISymbol, IOperation?> requiresPreviewFeaturesSymbols)
+        {
+            context.RegisterOperationAction(context => AnalyzeOperation(context.Operation, requiresPreviewFeaturesSymbols),
+                OperationKind.Invocation,
+                OperationKind.ObjectCreation,
+                OperationKind.PropertyReference,
+                OperationKind.FieldReference,
+                OperationKind.DelegateCreation,
+                OperationKind.EventReference
+                );
+
+            context.RegisterOperationBlockEndAction(context =>
+            {
+                if (thisAssemblyUsesPreviewFeatures)
+                {
+                    foreach (var symbolAndOperation in requiresPreviewFeaturesSymbols)
+                    {
+                        IOperation? value = symbolAndOperation.Value;
+                        if (value != null)
+                        {
+                            context.ReportDiagnostic(value.CreateDiagnostic(Rule));
+                        }
+                    }
```

Comment:
I'm not sure if this is related to the test issue. But could you try breaking after reporting a diagnostic in the loop.